---
title: Taxonomic Diversity and Distinctness Measures Applied to William's Lake Soil
  Metagenomes
author: "Niels Hanson, Aria Hahn"
date: "January 16, 2015"
output:
  html_document:
    keep_md: yes
    number_sections: yes
    theme: readable
    toc: yes
csl: ieee.csl
bibliography: taxonomic_variance_notes.bib
---

```{r echo=FALSE}
wd <- "~/Dropbox/projects/TaxonomicVariance"
setwd(wd)
```

# Summary

This document details the implementation of the Clarke's and Warwick's Diversity and Distinctness measures using the functional annotation to reference genomes [[@Warwick:1995vf, @Pienkowski:1998um, @Clarke:1999tx]]. However, the current implementation in the Vegan package does not take into account incomplete taxonomies [[@Oksanen:2007vs]], a common situation when using functional-genes to infer taxonomy via the Lowest Common Ancestor method in meta'omic samples [[@Huson:2007jl]]. Moreover, we would like to take advantage of properties and intuition built into the Weighted Taxonomic Distance (WTD) [[@Hanson:2014bz]], which creates a hard separation between distances across sub-trees. To accomplish this we implement the calculation of distances from partial taxonomies, allowing the efficient calculation of WTD matrices. Moreover, we implement a number of analytical utility scripts to format taxonomic lineages the MetaPathways pipeline to create compatible distance and sample abundance matrix. Using the William's Lake (WL) soil metagenomes as a test case, we using that both taxonomic diversity measures, DeltaStar and DeltaPlus, show a positive correlation with increasing depth.

This document shows the development of the method step-by-step. First it describes the original behavior of the implemented taxonomic diversity measures as implemented in the Vegan package, demonstrating their incompatibility with partial taxonomies. Next, we describe the implementation of a matrix form calculation of partial taxonomies, and through weighting demonstrate the implementation of the WTD. Next, we describe the development and use of utility scripts for use with MetaPathways output for creating taxonomies with respect to samples, pathways, and reactions, using presence/absence, ORF counts, or a normalized read-mapping measure, Reads per Kilobase per Million Mapped reads. Finally, we demonstrate the use of this implementation using 25 metagenomic samples from the William's Lake long term soil productivity (LTSP) site, and show that both diversity metrics reflect the increase in taxonomic diversity with increasing depth in the natural, undisturbed treatment, while a flat or negative trend is observed in the clear-cut treatment. 

## Contents

This analysis consists of the following files and data products:

* [scripts/](): Directory containing perl and python scripts to handle MetaPathways output and ePGDBs
     * [extract_pathway_table_from_pgdb.pl](): Slightly-modified version of the MetaPathways script to extract pathways in the 'long' format with pathways and reactions as variables
     * [link_refseq_taxa_to_pathway.py](): Script to link pathway tables with RefSeq Best-HIT, Lowest Common Ancestor (LCA), LCA* taxonomies, as well as RPKM values. This script requires the parsed B/LAST outputs, RPKM files, produced by MetaPathways, and the 'long' pathway tables produced by `extract_pathway_table_from_pgdb.pl` from the ePGDBs. Large tables can be gziped (.gz) for convenience.
     * [extract_LTSP_WL_pwys.sh](): Shell script uses `extract_pathway_table_from_pgdb.pl` to extract WL pathways from Pathway Tools
     * [run_link_refseq_taxa_to_pathway.sh](): Shell script runs `link_refseq_taxa_to_pathway.py` on the WL samples
     * [python_resources/](): Folder containing classes and methods for running `link_refseq_taxa_to_pathway.py`
          * [ncbi_taxonomy_tree.txt](): Slightly-modified NCBI Taxonomy Database Hierarchy.
          * [ncbi.map](): Preferred taxonomy names from NCBI IDs.
          * [LCAStar.py](): Class and functions for calculating LCA, MEGAN Taxonomy, and LCA*.

* [annotations/](): Directory containing RefSeq annotations (`.LASTout.parsed.txt.gz`), RPKM values (`.orf_rpkm.txt.gz`), and long-style pathway tables (`.long.pwy.txt.gz`) for the WL LTSP samples.
* [r_scripts/](): Directory of r functions.
     * [my_taxa2dist.R](): Modification of the `taxa2dist()` to handle partial taxonomies and the WTD
* [data/](): Processed data ready for analysis in R
     * [simple_test.txt](): Very basic taxonomy matrix with incomplete taxonomies for testing our implementation
     * [Phylodisttest.txt](): Matrix of partial taxonomies from the WL LTSP samples 
     * [LTSP_WL_MappingandInfo_clean.csv](): Matrix of metadata, chemistry and experimental conditions for the WL samples
          * [pwy_taxa/](): long-style pathway tables with LCA taxonomies and RPKM values from the LTSP samples
     * [temp/](): Directory to store previous `.Rdata` files for fast processing. Remove contents of this directory to reboot the analysis and reprocess taxonomic distance matrix
* [figures/](): Directory for pdf figures in the analysis.

## R Libraries

* Load R libraries

```{r message=FALSE, warning=FALSE}
library(vegan)
library(pheatmap)
library(dplyr)
library(ggplot2)
library(reshape2)
library(RColorBrewer)
figure_dir <- paste(wd, "figures", sep="/")
```

# Implementation

The initial implementation in VEGAN does allow for incomplete taxonomies. In this section we demonstrate this behavior using the Dutch Dune dataset and a subset of the WL LTSP partial taxonomies. The Dutch Dune Meadow dataset has full taxonomies and is not particularly large compared to the number of unique taxonomies found in environmental datasets. We observe that the current implementation of `taxa2dist`, while improved to some degree class-based scaling, produces distances that do not respect partial taxonomies.

* Load the Dutch Dune Meadow data.

```{r}
data(dune)
data(dune.taxon)
```

* Run through the example data

```{r}
# calculate the taxonomic distance
dune.taxdis <- taxa2dist(dune.taxon, varstep=TRUE)
# plot the tree
dune.taxontree <- hclust(dune.taxdis)
plot(dune.taxontree)
```

* Plotting the matrix as a heatmap will show the structure in more detail

```{r fig.width=10, fig.height=10}
pheatmap(dune.taxdis, treeheight_row=150, treeheight_col=150)
```

* Vegan's implementation of the taxonomic distance

```{r}
mod <- taxondive(dune, dune.taxdis)
summary(mod)
```

* Try it out with some partial taxonomies

```{r}
taxon.test = read.table(paste(wd, "data/Phylodisttest.txt", sep="/"), header=TRUE, row.names="row.names", sep ="\t")
```

* Build a tree based on the distances as calculated via the original implementation of taxa2dist in Vegan. We will vary the step lengths between successive levels relative to proportional loss of the number of distinct classes.

```{r fig.height=12}
taxon.test.matrix <- taxa2dist(taxon.test, varstep=TRUE)
taxontree <- hclust(taxon.test.matrix)
plot(taxontree)
```

* Structure with number of distinct class step lengths.

```{r fig.width=12, fig.height=12}
pheatmap(taxon.test.matrix, treeheight_row=150, treeheight_col=150)
pdf(file = paste(figure_dir, "WL_matrix_class_step.pdf", sep="/"), width=12, height=12)
pheatmap(as.matrix(taxon.test.matrix), treeheight_row=150, treeheight_col=150)
dev.off()
```

* Uniform step lengths

```{r fig.height=12}
taxon.test.matrix2 <- taxa2dist(taxon.test)
taxontree2 <- hclust(taxon.test.matrix2)
plot(taxontree2)
```

* Structure with uniform step lengths.

```{r fig.width=12, fig.height=12}
pheatmap(taxon.test.matrix2, treeheight_row=150, treeheight_col=150)
```

## Implementing WTD and Partial Taxonomy

This demonstrates our implementation of the weighted taxonomic distance and partial taxonomies in `my_taxa2dist.R`, a slightly modified version of the original. 

*TODO: Add some details of the matrix form exclusive OR and the outer product form of the taxonomic distance matrix.*

* source the implementation of `my_taxa2dist`:

```{r}
source(paste(wd, "r_scripts/my_taxa2dist.R", sep="/"))
```

* we'll do a simple test to do a sanity check of the implementation

```{r}
simple_test1 <- read.table(paste(wd, "data/simple_test.txt", sep="/"), header=TRUE, row.names="row.names", sep ="\t")
simple_test1_dist <- my_taxa2dist(simple_test1, check=FALSE, wtd=TRUE)
simple_test1_dist
```

* plot the simple tree

```{r}
plot(hclust(simple_test1_dist))
```

* We'll do another quick test with the LTSP data

```{r fig.width=9.960317, fig.height=9.614173}
taxon.test.wtd <- my_taxa2dist(taxon.test, check=FALSE, wtd=TRUE)
pheatmap(as.matrix(taxon.test.wtd), treeheight_row=150, treeheight_col=150)

# create pdf
pdf(file = paste(figure_dir, "WL_matrix_wtd.pdf", sep="/"), width=9.9603, height=9.614)
pheatmap(as.matrix(taxon.test.wtd), treeheight_row=150, treeheight_col=150)
dev.off()

```

* lets take a look at the Dune data again, looks about the same

```{r fig.width=16, fig.height=10}
# calculate the taxonomic distance
dune.taxdis.wtd <- my_taxa2dist(dune.taxon, check=FALSE, varstep=FALSE, wtd=TRUE)
# plot the tree
dune.taxontree.wtd <- hclust(dune.taxdis.wtd)
par(mfrow=c(1,2))
plot(dune.taxontree, main="Dune (Original)")
plot(dune.taxontree.wtd, main="Dune (WTD)" )
par(mfrow=c(1,1))

# create pdf
pdf(file = paste(figure_dir, "dune_wtd_compare.pdf", sep="/"), width=16, height=10)
par(mfrow=c(1,2))
plot(dune.taxontree, main="Dune (Original)")
plot(dune.taxontree.wtd, main="Dune (WTD)" )
par(mfrow=c(1,1))
dev.off()

mod <- taxondive(dune, dune.taxdis.wtd)
summary(mod)
```

## Taxonomy and Pathways from MetaPathways

* Slightly modified MetaPathways utility script `extract_pathway_table_from_pgdb.pl` that produces long-table output labeling pathways and reactions
* wrote `scripts/link_refseq_taxa_to_pwy.py` to combine output of pathways in metapathways with parsedblast and RPKM values

```
# Run Pathway Tools in API mode
./pathway-tools/pathway-tools -api

# navigate to the scripts directory 
cd .../scripts/

# Extract pathways
./extract_LTSP_WL_pwys.sh

# Calculate LCA and prepare pathway taxonomies
./run_link_refseq_taxa_to_pathway.sh
```

# Analysis of William's Lake

Here we utilize our implementation to analyze the William's Lake metagenomes by sample and pathway.

## Diversity by Sample

In order to calculate the diversity measures, we need to create a distance matrix and a sample by taxa abundance matrix. To do this we will parse all the extract taxonomies and create a non-redundant list of taxonomies, and then use these to create the distance matrix for our samples using the WTD. From here we construct two sample by taxa abundance matrices; the first based on the raw ORF counts, and the second based on the RPKM normalized abundance measure. In both cases we show that the taxonomic diversity and distinctness measures, delta* and delta+, show a strong positive correlation with depth in the Natural samples, while the treated samples show no clear or negative correlations.

* Various functions for creating properly formatted taxa matrix from partial taxonomies for `my_taxa2dist()`

```{r}
# fills in the rest of the NA
create_rows <- function(x, max) {
  c(x, rep(NA, max - length(x)))
}

# creates dataframe delimiting the taxonomy hierarchy
create_taxa_df <- function(x) {
  split_taxa <- sapply(x, strsplit, ";")
  lengths <- sapply(split_taxa, length)
  max_length <- max(lengths)
  rows <- lapply(split_taxa, create_rows, max_length)
  rows <- lapply(rows, rev) # reverse rows
  taxa_df <- t(data.frame(rows))
  row.names(taxa_df) <- x
  taxa_df[is.na(taxa_df)] <- "unclassified"
  return(taxa_df)
}
```

* Create a non-redundant list of unique taxonomies for each of the annotations

```{r cache=TRUE}
pwy_taxa_dir <- paste(wd, "data/pwy_taxa/", sep="/")
temp_dir <- paste(wd, "data", "temp", sep="/")

# select files from the pwy_taxa_dir
pwy_taxa_files <- list.files(pwy_taxa_dir, pattern="*taxa.txt.gz")
rdata_files <- list.files(temp_dir, pattern="*.Rdata")

# create non-redundant list of taxonomies for all samples
nr_taxa_list <- c()
if(!("nr_taxa_list.Rdata" %in% rdata_files)) { 
  for (pt_file in pwy_taxa_files) {
    print(pt_file)
    df <- read.table(paste(pwy_taxa_dir, pt_file, sep="/"), header=TRUE, sep="\t",
                   colClasses= c('character', 'character', 'character', 'character', 'character',
                                 'numeric', 'numeric', 'numeric', 'character', 'character', 'numeric'), quote = "")
    my_taxa <- unique(df$TAXONOMY)
    nr_taxa_list <- c(nr_taxa_list, my_taxa)
  }
  # save to disk
  nr_taxa_list <- unique(nr_taxa_list)
  save(nr_taxa_list, file=paste(temp_dir, "nr_taxa_list.Rdata", sep="/"))
} else {
  load(paste(temp_dir, "nr_taxa_list.Rdata", sep="/"))
}
# remove NA class
nr_taxa_list <- nr_taxa_list[!is.na(nr_taxa_list)]
```

* Construct the WTD distance matrix from these annotations

```{r cache=TRUE}
if(!("taxdis.Rdata" %in% rdata_files)) { 
  nr_taxa_list <- unique(nr_taxa_list)
  split_taxa <- sapply(nr_taxa_list, strsplit, ";")
  lengths <- sapply(split_taxa, length)
  max_length <- max(lengths)
  test_df <- create_taxa_df(nr_taxa_list)
  taxdis <- my_taxa2dist(test_df, check=FALSE, varstep=FALSE, wtd=TRUE)

  # save to disk
  save(taxdis, file=paste(temp_dir, "taxdis.Rdata", sep="/"))
} else {
  load(paste(temp_dir, "taxdis.Rdata", sep="/"))
}
```

* Construct the sample by taxa matrix (ORF counts)

```{r}
sample_taxa_matrix_orfs <- NULL
if(!("sample_taxa_matrix_orfs.Rdata" %in% rdata_files)) { 
  for (pt_file in pwy_taxa_files) {
    print(pt_file)
    df <- read.table(paste(pwy_taxa_dir, pt_file, sep="/"), header=TRUE, sep="\t",
                    colClasses= c('character', 'character', 'character', 'character', 'character',
                                  'numeric', 'numeric', 'numeric', 'character', 'character', 'numeric'), quote = "")
    my_row <- table(df$TAXONOMY)[nr_taxa_list]
    my_row[is.na(my_row)] <- 0
    sample_taxa_matrix_orfs <- rbind(sample_taxa_matrix_orfs,my_row)
  }
  # set sample as row_names
  colnames(sample_taxa_matrix_orfs) <- nr_taxa_list
  row.names(sample_taxa_matrix_orfs) <- sub("-scaffolds.long.pwy.taxa.txt.gz", "", pwy_taxa_files)
  
  save(sample_taxa_matrix_orfs, file=paste(temp_dir, "sample_taxa_matrix_orfs.Rdata", sep="/"))
} else {
  load(paste(temp_dir, "sample_taxa_matrix_orfs.Rdata", sep="/"))
}
```

* Calculate the diversity scores (ORF Counts)

```{r}
if(!("diversity_result_orfs.Rdata" %in% rdata_files)) { 
  diversity_result_orfs <- taxondive(sample_taxa_matrix_orfs, taxdis)
  save(diversity_result_orfs, file=paste(temp_dir, "diversity_result_orfs.Rdata", sep="/"))
} else {
  load(paste(temp_dir, "diversity_result_orfs.Rdata", sep="/"))
}
div_res_orfs <- summary(diversity_result_orfs)
div_res_orfs
```

* Load the William's Lake metadata

```{r}
wl_meta_df <- read.table(paste(wd, "data", "LTSP_WL_MappingandInfo_clean.csv", sep="/"), sep=",", header=TRUE)
```

* Create summary Data Frame (ORF Counts)

```{r}
div_res_orfs_matrix <- as.matrix(div_res_orfs)
div_res_orfs_matrix <- cbind(div_res_orfs_matrix[-nrow(div_res_orfs_matrix),], as.matrix(wl_meta_df)) # combine with df
master_df_orfs <- data.frame(delta=as.numeric(div_res_orfs_matrix[,1]),
           delta_star=as.numeric(div_res_orfs_matrix[,2]),
           delta_plus=as.numeric(div_res_orfs_matrix[,3]),
           delta_plus_sd=as.numeric(div_res_orfs_matrix[,4]),
           delta_plus_z=as.numeric(div_res_orfs_matrix[,5]),
           delta_plus_pr=as.numeric(div_res_orfs_matrix[,6]),
           horizon=as.character(div_res_orfs_matrix[,8]),
           treatment=as.character(div_res_orfs_matrix[,9]),
           depth=as.numeric(div_res_orfs_matrix[,10]),
           pH=as.numeric(div_res_orfs_matrix[,11]),
           totalC_percent=as.numeric(div_res_orfs_matrix[,12]),
           totalN_percent=as.numeric(div_res_orfs_matrix[,13]),
           totalorganicN_ppm=as.numeric(div_res_orfs_matrix[,14]),
           nh4_ppm=as.numeric(div_res_orfs_matrix[,15]),
           no3_ppm=as.numeric(div_res_orfs_matrix[,16]))
```

* Depth (cm) by taxonomic diversity (ORF Counts)

```{r fig.width=9.832, fig.height=4.399}
p <- ggplot(master_df_orfs, aes(x=depth, y=delta_star)) 
p <- p + geom_smooth(method="glm", size=1.5) 
p <- p + geom_point(size=3.5, aes(col=horizon)) 
p <- p + theme_bw() 
p <- p + facet_wrap(~treatment)
p <- p + ggtitle("Williams Lake (ORF Counts)")
p
pdf(file = paste(figure_dir, "wl_delta_star_orfs.pdf", sep="/"), width=9.832, height=4.399)
p
dev.off()
```

```{r fig.width=9.832, fig.height=4.399}
# delta plus (ORF count)
# quartz(width=9.832677, height=4.399606)
p <- ggplot(master_df_orfs, aes(x=depth, y=delta_plus)) 
p <- p + geom_smooth(formula=y ~ x, method="glm", size=1.5) 
p <- p + geom_point(size=3.5, aes(col=horizon)) 
p <- p + facet_wrap(~treatment) 
p <- p + theme_bw()
p <- p + ggtitle("Williams Lake (ORF Counts)")
p
pdf(file = paste(figure_dir, "wl_delta_plus_orfs.pdf", sep="/"), width=9.832, height=4.399)
p
dev.off()
```

* Construct the sample by taxa matrix (RPKM)

```{r}
div_res_rpkm_matrix <- NULL
if(!("div_res_rpkm_matrix.Rdata" %in% rdata_files)) { 
  for (pt_file in pwy_taxa_files) {
    print(pt_file)
    df <- read.table(paste(pwy_taxa_dir, pt_file, sep="/"), header=TRUE, sep="\t",
                   colClasses= c('character', 'character', 'character', 'character', 'character',
                                 'numeric', 'numeric', 'numeric', 'character', 'character', 'numeric'), quote = "")
    
    res <- df %>% 
    select(TAXONOMY, RPKM) %>%
    group_by(TAXONOMY) %>% 
    summarize( rpkm=sum(RPKM) )
    
    my_row <- res$rpkm
    names(my_row) <- res$TAXONOMY
    my_row <- my_row[nr_taxa_list]
    my_row[is.na(my_row)] <- 0
    div_res_rpkm_matrix <- rbind(div_res_rpkm_matrix,my_row)
  }
  
  # set sample as row_names
  colnames(div_res_rpkm_matrix) <- nr_taxa_list
  row.names(div_res_rpkm_matrix) <- sub("-scaffolds.long.pwy.taxa.txt.gz", "", pwy_taxa_files)
  
  save(div_res_rpkm_matrix, file=paste(temp_dir, "div_res_rpkm_matrix.Rdata", sep="/"))
} else {
  load(paste(temp_dir, "div_res_rpkm_matrix.Rdata", sep="/"))
}
```

* RPKM diversity results

```{r}
if(!("diversity_result_rpkm.Rdata" %in% rdata_files)) { 
  diversity_result_rpkm <- taxondive(div_res_rpkm_matrix, taxdis)
  save(diversity_result_rpkm, file=paste(temp_dir, "diversity_result_rpkm.Rdata", sep="/"))
} else {
  load(paste(temp_dir, "diversity_result_rpkm.Rdata", sep="/"))
}
div_res_rpkm <- summary(diversity_result_rpkm)
div_res_rpkm
```

* RPKM result matrix

```{r}
res_matrix_rpkm <- as.matrix(div_res_rpkm)
res_matrix_rpkm <- cbind(res_matrix_rpkm[-nrow(res_matrix_rpkm),], as.matrix(wl_meta_df)) # combine with df
master_df_rpkm <- data.frame(delta=as.numeric(res_matrix_rpkm[,1]),
           delta_star=as.numeric(res_matrix_rpkm[,2]),
           delta_plus=as.numeric(res_matrix_rpkm[,3]),
           delta_plus_sd=as.numeric(res_matrix_rpkm[,4]),
           delta_plus_z=as.numeric(res_matrix_rpkm[,5]),
           delta_plus_pr=as.numeric(res_matrix_rpkm[,6]),
           horizon=as.character(res_matrix_rpkm[,8]),
           treatment=as.character(res_matrix_rpkm[,9]),
           depth=as.numeric(res_matrix_rpkm[,10]),
           pH=as.numeric(res_matrix_rpkm[,11]),
           totalC_percent=as.numeric(res_matrix_rpkm[,12]),
           totalN_percent=as.numeric(res_matrix_rpkm[,13]),
           totalorganicN_ppm=as.numeric(res_matrix_rpkm[,14]),
           nh4_ppm=as.numeric(res_matrix_rpkm[,15]),
           no3_ppm=as.numeric(res_matrix_rpkm[,16]))
```

* Depth by taxonomic diversity (RPKM)

```{r fig.width=9.832, fig.height=4.399}
# delta_star (RPKM count)
p <- ggplot(master_df_rpkm, aes(x=depth, y=delta_star)) 
p <- p + geom_smooth(method="glm", size=1.5) 
p <- p + geom_point(size=3.5, aes(col=horizon)) 
p <- p + theme_bw() 
p <- p + facet_wrap(~treatment)
p <- p + ggtitle("William's Lake (RPKM)")
p
pdf(file = paste(figure_dir, "wl_delta_star_rpkm.pdf", sep="/"), width=9.832, height=4.399)
p
dev.off()
```

```{r fig.width=9.832, height=4.399}
# delta plus (RPKM count)
p <- ggplot(master_df_rpkm, aes(x=depth, y=delta_plus)) 
p <- p + geom_smooth(formula=y ~ x, method="glm", size=1.5) 
p <- p + geom_point(size=3.5, aes(col=horizon)) 
p <- p + facet_wrap(~treatment) 
p <- p + theme_bw()
p <- p + ggtitle("William's Lake (RPKM)")
p
pdf(file = paste(figure_dir, "wl_delta_plus_rpkm.pdf", sep="/"), width=9.832, height=4.399)
p
dev.off()
```

## Diversity by Pathway

* Construct the Sample, Pathway, Measure, and Values table 
     * Calculate diversity per pathway

```{r}
# cleans up summary results from taxondive
construct_summary_matrix <- function(taxon_dive_summary, clean=TRUE) {
  x <- data.frame(delta=taxon_dive_summary[,1],
                 delta_star=taxon_dive_summary[,2],
                 delta_plus=taxon_dive_summary[,3],
                 delta_plus_sd=taxon_dive_summary[,4],
                 delta_plus_z=taxon_dive_summary[,5],
                 delta_plus_pr=taxon_dive_summary[,6])
  # save expected values
  delta_e <- x[nrow(x),1]
  delta_star_e <- x[nrow(x),2]
  delta_plus_e <- x[nrow(x),3]
  
  x$PWY_NAME = rownames(x)
  if (clean) {
    # removes NAs, NaN, Inf
    x <- x[complete.cases(x),]
    # must be positive 
    positive_set <- (x[,1] >= 0 & x[,2] >= 0 & x[,3] >= 0)
    x <- x[positive_set,]
  }
  
  # bind results together
  res <- list(df=x, delta_e=delta_e, delta_star_e=delta_star_e, delta_plus_e=delta_plus_e)
  return(res)
}
```

```{r}
load(paste(temp_dir, "pathway_master_table.Rdata", sep="/"))
# clean up names
pathway_master_table$SAMPLE <- sub("-scaffolds", "", pathway_master_table$SAMPLE)
pathway_master_table$SAMPLE <- sub("a", "A", pathway_master_table$SAMPLE)
```
```

```{r eval=FALSE}
if("pathway_master_table.Rdata" %in% rdata_files) {
  load(paste(temp_dir, "pathway_master_table.Rdata", sep="/"))
} else {
  pathway_master_table = NULL
  for (pt_file in pwy_taxa_files) {
    print(pt_file)
    # load dataframe
    df <- read.table(paste(pwy_taxa_dir, pt_file, sep="/"), header=TRUE, sep="\t",
                    colClasses= c('character', 'character', 'character', 'character', 'character',
                                 'numeric', 'numeric', 'numeric', 'character', 'character', 'numeric'), quote = "")
    ## calculate diversity per pathway
    
    # summarize pathways by taxonomy abundance
    res <- df %>% 
        select(SAMPLE, PWY_NAME, PWY_COMMON_NAME, TAXONOMY, NUM_REACTIONS, NUM_COVERED_REACTIONS, ORF_COUNT, RPKM) %>%
        group_by(SAMPLE, PWY_NAME, PWY_COMMON_NAME, TAXONOMY) %>%
        filter( is.na(TAXONOMY) != TRUE) %>%
        summarize( ORF_COUNT=max(ORF_COUNT), RPKM=sum(RPKM)  )
    
    # convert into dataframe for taxonomic calculation
    res2 <- dcast(res, PWY_NAME ~ TAXONOMY, value.var='RPKM', sum)
    rownames(res2) <- res2$PWY_NAME
    res2 <- res2[-1]

    # prepare taxonomic distance matrix
    my_taxa_list <- unique(colnames(res2))
    split_taxa <- sapply(my_taxa_list, strsplit, ";")
    lengths <- sapply(split_taxa, length)
    max_length <- max(lengths)
    test_df <- create_taxa_df(my_taxa_list)
    my_taxdis <- my_taxa2dist(test_df, check=FALSE, varstep=FALSE, wtd=TRUE)
    
    # calculate statistics
    test <- taxondive(res2, my_taxdis)
    summary_res <- summary(test)
    
    # clean up summary
    test_list <- construct_summary_matrix(summary_res)

    # create a data frame of sample, pathway, measure, value
    res3 <- df %>% 
        select(SAMPLE, PWY_NAME, PWY_COMMON_NAME, TAXONOMY, NUM_REACTIONS, NUM_COVERED_REACTIONS, ORF_COUNT, RPKM) %>%
        group_by(SAMPLE, PWY_NAME, PWY_COMMON_NAME) %>%
        filter( is.na(TAXONOMY) != TRUE) %>%
        summarize( ORF_COUNT=max(ORF_COUNT), RPKM=sum(RPKM)  )

    # bind new statistics to pathway
    res4 <- merge(res3, test_list$df, by="PWY_NAME")
    
    # add to master table
    pathway_master_table <- rbind(pathway_master_table, res4)
  }
  
  # save when complete
  save(pathway_master_table, file=paste(temp_dir, "pathway_master_table.Rdata", sep="/"))
}

# clean up names
pathway_master_table$SAMPLE <- sub("-scaffolds", "", pathway_master_table$SAMPLE)
pathway_master_table$SAMPLE <- sub("a", "A", pathway_master_table$SAMPLE)
```

* Now with the pathway master table constructed we can compute a few things

```{r echo=FALSE, eval=FALSE}
p <- ggplot(pathway_master_table, aes(x=ORF_COUNT, y=RPKM)) + geom_point() + facet_wrap(~SAMPLE)
p
```

```{r echo=FALSE, eval=FALSE}
p <- ggplot(pathway_master_table, aes(x=delta_star)) + geom_histogram() + facet_wrap(~SAMPLE)
p
```

```{r eval=FALSE}
# join by metadata
names(wl_meta_df)[names(wl_meta_df) == 'Library'] <- "SAMPLE"
pathway_master_table <- merge(pathway_master_table, wl_meta_df, "SAMPLE")

res<-pathway_master_table %>%
  select(SAMPLE, PWY_NAME, PWY_COMMON_NAME, ORF_COUNT, delta_star) %>%
  group_by(SAMPLE, PWY_NAME) %>%
  summarize(PWY_COMMON_NAME, min_delta_star = min(delta_star, na.rm = TRUE), ORF_COUNT) %>%
  filter(ORF_COUNT > 10) %>%
  filter(min_rank(min_delta_star) < 5)
```

# Future Ideas

Ideas for future work.

* Bootstrap p-values for delta_star (could be time-consuming)
* Investigate measures in the context of the pathway

# References

Section for work cited.
